File: python/com/nari/algorithm/_102_binary_tree_level_order_traversal.py
"""
* 题目地址：https://leetcode-cn.com/problems/binary-tree-level-order-traversal/
* 题目概述：Given the root of a binary tree, return the level order traversal of
*         its nodes' values. (i.e., from left to right, level by level).

* 来源：力扣（LeetCode）
* 链接：https://leetcode-cn.com/problems/3sum
* 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
"""

from collections import deque
from com.nari.algorithm.model.Node import Node


def level_order(root):
    if not root:
        return []

    result = []  # 二维数组
    queue = deque()  # 双端队列，存放当前层的节点
    queue.append(root)  # 根节点对象（第一层）首先入列

    while queue:
        current_level_size = len(queue)  # 当前层节点数目，为下文for循环的次数
        current_level_node_vals = []  # 存放当前层节点的值
        for _ in range(current_level_size):
            node = queue.popleft()  # 始终从队列左端弹出本层的节点
            current_level_node_vals.append(node.val)
            if node.left:
                queue.append(node.left)  # 下一层的新节点放在队列尾部
            if node.right:
                queue.append(node.right)  # 下一层的新节点放在队列尾部
        result.append(current_level_node_vals)

    return result


def main():

    #       23
    #    5       7
    #  2   9   4   3
    node31 = Node(2)
    node32 = Node(9)
    node33 = Node(4)
    node34 = Node(3)
    node21 = Node(5, node31, node32)
    node22 = Node(7, node33, node34)
    node11 = Node(23, node21, node22)
    print(level_order(node11))


if __name__ == '__main__':
    main()

File: java/src/com/nari/algorithm/_102_BinaryTreeLevelOrderTraversal.java
package com.nari.algorithm;

import com.nari.algorithm.model.TreeNode;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/binary-tree-level-order-traversal/
 */
public class _102_BinaryTreeLevelOrderTraversal {

    public static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) return res;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while(!queue.isEmpty()) {
            int current_level_size = queue.size();
            List<Integer> current_node_vals = new ArrayList<>();
            for (int i = 0; i < current_level_size; i++) {
                TreeNode currNode = queue.poll();
                assert currNode != null;
                current_node_vals.add(currNode.getVal());
                if (currNode.getLeft() != null) {
                    queue.add(currNode.getLeft());
                }
                if (currNode.getRight() != null) {
                    queue.add(currNode.getRight());
                }
            }
            res.add(current_node_vals);
        }

        return res;
    }

    public static void main(String[] args) {
        //       23
        //    5       7
        // 2   9   4   3
        TreeNode node31 = new TreeNode(2);
        TreeNode node32 = new TreeNode(9);
        TreeNode node33 = new TreeNode(4);
        TreeNode node34 = new TreeNode(3);
        TreeNode node21 = new TreeNode(5, node31, node32);
        TreeNode node22 = new TreeNode(7, node33, node34);
        TreeNode node11 = new TreeNode(23, node21, node22);
        System.out.println(levelOrder(node11));
    }

}
